sympy/polys/matrices/domainmatrix.py

```diff
@@ -1692,6 +1800,601 @@ def det(self):
+        instead. If the intention is to compute the inverse matrix or solve a
+        system of equations then :meth:`inv_den` is more efficient.
+
+        Examples
+        ========
+
+        >>> from sympy import ZZ
+        >>> from sympy.polys.matrices import DomainMatrix
+        >>> A = DomainMatrix([
+        ...     [ZZ(2), ZZ(-1), ZZ(0)],
+        ...     [ZZ(-1), ZZ(2), ZZ(-1)],
+        ...     [ZZ(0), ZZ(0), ZZ(2)]], (3, 3), ZZ)
+        >>> Ainv, den = A.inv_den()
+        >>> den
+        6
+        >>> Ainv
+        DomainMatrix([[4, 2, 1], [2, 4, 2], [0, 0, 3]], (3, 3), ZZ)
+        >>> A * Ainv == den * A.I().to_dense()
+        True
+
+        Parameters
+        ==========
+
+        method : str, optional
+            The method to use to compute the inverse. Can be one of ``None``,
+            ``'rref'`` or ``'charpoly'``. If ``None`` then the method is
+            chosen automatically (see :meth:`solve_den` for details).
+
+        See Also
+        ========
+
+        inv
+        det
+        adj_det
+        solve_den
+        """
+        return self.solve_den(self.I(), method=method)
+
+    def solve_den(self, b, method=None):
+        """
+        Solve matrix equation $Ax = b$ without fractions in the ground domain.
+
+        Examples
+        ========
+
+        Solve a matrix equation over the integers:
+
+        >>> from sympy import ZZ
+        >>> from sympy.polys.matrices import DM
+        >>> A = DM([[ZZ(1), ZZ(2)], [ZZ(3), ZZ(4)]], ZZ)
+        >>> b = DM([[ZZ(5)], [ZZ(6)]], ZZ)
+        >>> xnum, xden = A.solve_den(b)
+        >>> xden
+        -2
+        >>> xnum
+        DomainMatrix([[8], [-9]], (2, 1), ZZ)
+        >>> A * xnum == xden * b
+        True
+
+        Solve a matrix equation over a polynomial ring:
+
+        >>> from sympy import ZZ, symbols
+        >>> a, b, c, d, e, f = symbols('a b c d e f')
+        >>> R = ZZ[a, b, c, d, e, f]
+        >>> M = DM([[a, b], [c, d]], R)
+        >>> b = DM([[e], [f]], R)
+        >>> M.to_Matrix()
+        Matrix([
+        [a, b],
+        [c, d]])
+        >>> b.to_Matrix()
+        Matrix([
+        [e],
+        [f]])
+        >>> xnum, xden = M.solve_den_charpoly(b)
+        >>> xden
+        a*d - b*c
+        >>> xnum.to_Matrix()
+        Matrix([
+        [-b*f + d*e],
+        [ a*f - c*e]])
+        >>> M * xnum == xden * b
+        True
+
+        The solution can be expressed over a fraction field which will cancel
+        any gcd between the denominator and the elements of the numerator:
+
+        >>> xsol = xnum.to_field() / xden
+        >>> xsol.to_Matrix()
+        Matrix([
+        [(-b*f + d*e)/(a*d - b*c)],
+        [ (a*f - c*e)/(a*d - b*c)]])
+        >>> (M * xsol).to_Matrix() == b.to_Matrix()
+        True
+
+        The solution can also be expressed as a ``Matrix`` without attempting
+        any cancellation between the numerator and denominator:
+
+        >>> xnum.to_Matrix() / xnum.domain.to_sympy(xden)
```

Comment:
It would be nice if this demonstration gave a different result than the last one (i.e. with something uncancelled).

Reply:
Good idea. I've changed the example.